add copy subnet action for banned peer

ShaMan239 opened the pull request:
This PR adds a Copy Subnet context menu action to the Banned Peers Table.

This feature is quite helpful if a node using GUI might want to alert its peer about a particular malicious user. So it can copy that user’s subnet and broadcast it to its peers so they can ban it instantly using the setban command in the console.

| Master        | PR               |
| ----------- | ----------- |
| ![Screenshot_from_2021-07-21_00-01-331](https://user-images.githubusercontent.com/23396902/126377808-bd23bb19-3f47-4f1b-8371-39baa9747bbe.png) | ![Screenshot_from_2021-07-20_23-59-271](https://user-images.githubusercontent.com/23396902/126377824-818e0c9b-0058-481a-84eb-c8177ead69ce.png) |

hebasto commented:
Concept ACK.

> This PR adds a Copy Subnet context menu action to the Banned Peers Table.

But on the screenshot a menu item is called "Copy Address". Which variant is correct?

jarolrod commented on src/qt/rpcconsole.cpp:
```diff
@@ -706,6 +706,11 @@ void RPCConsole::setClientModel(ClientModel *model, int bestblock_height, int64_
         // create ban table context menu
         banTableContextMenu = new QMenu(this);
+        //: Context menu action to copy the subnet of a banned peer
+        banTableContextMenu->addAction(tr("Copy subnet"), [this] {
```
Since #362 we are now defaulting to adding mnemonic shortcuts for the context menus. This means that you should add one here

Below is how you would do so here:
```suggestion
        banTableContextMenu->addAction(tr("&Copy subnet"), [this] {
```

ShaMan239 replied:
this is addressed in d518d67

jarolrod reviewed (commented):
Concept ACK

Looking good, one note: https://github.com/bitcoin-core/gui/pull/384/files#r673396366

hebasto commented:
@ShaMan239

And warm welcome as a new contributor :fireworks:

ShaMan239 commented:
> @ShaMan239
> 
> And warm welcome as a new contributor fireworks

Thank You very much @hebasto

kristapsk commented:
Concept ACK. I think "Copy address" should be added to connected peers list's context menu too, both for consistency and usefulness.

RandyMcMillan commented:
Concept ACK 

I will be curious to see how "Copy subnet" translates.
"Copy address" or simply "Copy" will be ok if there is a translation issue.

I lean toward simply "Copy".

It would be great to link a key command to the action as well - such as <control> + c maybe?

jarolrod commented:
@RandyMcMillan 
> I will be curious to see how "Copy subnet" translates.

That's a valid concern. Maybe we can add additional context to the translator comment stating that the subnet is a combination of its `IP address` and it's `Netmask`. Additionally we can add that if there is no clear translation in the target language the translation of `IP address` can be used.

hebasto commented:
> @RandyMcMillan
> 
> > I will be curious to see how "Copy subnet" translates.
> 
> That's a valid concern. Maybe we can add additional context to the translator comment stating that the subnet is a combination of its `IP address` and it's `Netmask`. Additionally we can add that if there is no clear translation in the target language the translation of `IP address` can be used.

If the translator comment will be touched, please, end it with a full stop.
Rationale: it is presented to translators in Transifex.com web-interface, and they usually do not look into the source code. The absence of a full stop could make an impression that Transifex.com just trimmed out a part of a comment.

ShaMan239 commented:
updated from 7c21de2 -> 32b4a54 ([pr384.02](https://github.com/ShaMan239/gui/tree/pr384.02) -> [pr384.03](https://github.com/ShaMan239/gui/tree/pr384.03))
changes: addressed https://github.com/bitcoin-core/gui/pull/384#issuecomment-884352943 and https://github.com/bitcoin-core/gui/pull/384#issuecomment-884357068

jarolrod reviewed (commented):
tACK 32b4a54429278dcfb6be1fbaa10c2b91e73a9a1a

Tested on macOS 11.3. Banned a peer, then was able to successfully copy the Peer's subnet by using the new `Copy subnet` context menu action.

The new translator comment looks good to me. I think the additional context behind it's meaning and a possible alternative will help with any translator confusion.

RandyMcMillan commented:
tACK 32b4a54

Tested on macOS 10.15.7

psancheti110 commented:
Tested 32b4a54429278dcfb6be1fbaa10c2b91e73a9a1a

I tried to build the PR on `macOS 11.4`. It delivers what is projected in the attached screenshots. The code patch looks neat and consistent. Congratulations on that 🎉 

Although I had two doubts about the functionality and the use-case mentioned for the added feature. 

1) The functionality implemented with the patch does not seem a new feature, instead just another way to copy the address. The Qt widget `banlistWidget` by default supports keyboard shortcuts like `Ctrl+C` on Linux/Windows and `⌘+C` on macOS. You could look at this comment for a better understanding of what I am trying to say: https://github.com/bitcoin-core/gui/issues/262#issuecomment-887350423

2) The use-case mentioned states about a node using GUI who might want to alert its peer about a particular malicious user. Would other nodes really welcome/interest such messages about a fellow peer, where they know nothing else about the user but its IP Address and assume it to be malicious by **Trusting** some other user?

In case I understood the use-case wrongly, would appreciate it if you'd correct me. 

Happy Coding!

ShaMan239 commented:
@psancheti110

Thank you for addressing these doubts. Let me try to answer them one by one:

1. Though we can directly copy addresses using keyboard shortcuts, this PR makes it more accessible for those who need to use the mouse.
2. Here is an example of when this is useful:
    - Sometimes, there is a list of peers known to spy that is passed around. This makes it more convenient to add to the list.
    - You have two GUI nodes and banned a peer on one node. You can copy the subnet from GUI Node A, take it to GUI Node B, and ban the peer there as well.

psancheti110 commented:
Hello @ShaMan239 

Appreciate you clearing my doubts and concerns 🙌🏻

1) Agreed. It can be a good addition for people who want to use a mouse for a change. 

I'd like to suggest something maybe you could add to make the use-case mentioned even better. 

- Currently, in the banned peers' list, it is not possible to select multiple addresses and copy them together. For now one has to copy individual addresses one by one.

- You can also checkout **rpcconsole**. Try running `listbanned` and that shall return a list of all banned peers in JSON format, which can be copied with just one click of the mouse. Attaching a screenshot for your reference. 
   > rpcconsole : `listbanned`

![Screenshot 2021-08-02 at 3 50 41 PM (2)](https://user-images.githubusercontent.com/54016434/127846678-aa99551c-6ae8-4f4c-bdc2-d54cee9b4501.png)

hebasto reviewed (commented):
> I think "Copy address" should be added to connected peers list's context menu too, both for consistency and usefulness.

Yes, I guess it is an improvement in terms of accessibility (that is why I concept ACKed this pr).

OTOH, such changes open a door for prs that suggest to add a "Copy some_another_fancy_column_content" into the context menus. It could ends with ugly bloated menus.

---

I don't think the use cases discussed above are a good practice for the following reasons:
- for an attacker it is easy and cheap to switch to another IP address in another subnet, and it seems unfeasible to track attacker manually
- aggressive banning (I believe it is the case when we use `setban` RPC call) increases the risk of splitting the network

jarolrod commented:
> Should "XXX" in "Copy XXX" menu action match a header title of the column?

I think that in the scope of this PR it should. While the combination of IP and Netmask is called subnet in other parts of the code, the column header here is 'IP/Netmask` and not `Subnet`. @hebasto is completely correct here. Having different names for the same data here can be confusing for both users and translators. 

For now lets keep the same name as in the header. If there is valid motivation to rename this header to `subnet`, then a follow up PR can do so and update the name of this action from `Copy IP/Netmask` to `Copy subnet`.

For now lets rename this action to `Copy IP/Netmask`

ShaMan239 commented:
updated from  32b4a54 ->  2da5d46 ([pr384.03](https://github.com/ShaMan239/gui/tree/pr384.03) -> [pr384.04](https://github.com/ShaMan239/gui/tree/pr384.04))
changes: addressed the comments [1](https://github.com/bitcoin-core/gui/pull/384#discussion_r682461249) and [2](https://github.com/bitcoin-core/gui/pull/384#issuecomment-902406496) by @hebasto and @jarolrod 
- Updated Name of Action from `copy subnet` to `copy IP/Netmask`
- Updated Translator's Comments accordingly

jarolrod commented on src/qt/rpcconsole.cpp:
```diff
@@ -706,6 +706,14 @@ void RPCConsole::setClientModel(ClientModel *model, int bestblock_height, int64_
         banTableContextMenu = new QMenu(this);
+        /*: Context menu action to copy the IP/Netmask of a banned peer.
+            IP/Netmask is the combination of a peer's IP address and it's Netmask.
+            For IP Address see: https://en.wikipedia.org/wiki/Subnetwork
```
This is similar to what was done in #332, see: https://github.com/bitcoin-core/gui/pull/332/files#diff-fe41db46679f280131a7b0a04d39383dd4d0ab623cec8de8ddc74f79b163dabfR285

But, is this an appropriate link for more information on `IP Address`. What about: https://en.wikipedia.org/wiki/IP_address

ShaMan239 commented:
Updated from 2da5d46 to a52f72f
Addressed [this](https://github.com/bitcoin-core/gui/pull/384#discussion_r692998313) comment by @jarolrod
- Updated translator comment to now direct towards the correct link with the information regarding the meaning of IP Address.

jarolrod commented:
tACK a52f72f

tested on Ubuntu 20.04 and macOS 12. PR works as intended. Translator comments look good to me

hebasto commented on src/qt/rpcconsole.cpp:
```diff
@@ -706,6 +706,14 @@ void RPCConsole::setClientModel(ClientModel *model, int bestblock_height, int64_
+        /*: Context menu action to copy the IP/Netmask of a banned peer.
+            IP/Netmask is the combination of a peer's IP address and it's Netmask.
+            For IP Address see: https://en.wikipedia.org/wiki/IP_address
+            For NetMask see: https://wiki.teltonika-networks.com/view/What_is_a_Netmask%3F */
```
I'd avoid mentioning any links to proprietary sites in the Bitcoin Core codebase.

Maybe, just skip last two lines of the comment?

jarolrod replied:
wikipedia or the other website? we already link to wikipedia articles to help with translations: https://github.com/bitcoin-core/gui/blob/d3203a99d886177eee9d1f9cd8411e215118a4e6/src/qt/addressbookpage.cpp#L285

hebasto replied:
At least, drop `https://wiki.teltonika-networks.com`, please.
